File: scripts/tt_try_fetch.py
from __future__ import annotations

import asyncio
import json
import os
import sys
from datetime import datetime, timezone
from typing import Any
from urllib import request as urlrequest

from TikTokApi import TikTokApi


def parse_sound_ids(raw: str) -> list[str]:
    return [part.strip() for part in raw.split(",") if part.strip()]


def to_int(value: Any) -> int:
    try:
        if isinstance(value, str):
            return int(value.replace(",", "").strip())
        return int(value)
    except Exception:
        return 0


def to_bool(value: Any, default: bool = False) -> bool:
    if value is None:
        return default
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"1", "true", "yes", "y", "on"}:
            return True
        if normalized in {"0", "false", "no", "n", "off"}:
            return False
    return default


def to_iso(value: Any) -> str:
    if isinstance(value, (int, float)):
        return datetime.fromtimestamp(int(value), tz=timezone.utc).isoformat()
    if isinstance(value, str):
        return value
    return datetime.now(tz=timezone.utc).isoformat()


def post_to_function(function_url: str, bearer_token: str, payload: dict[str, Any]) -> dict[str, Any]:
    body = json.dumps(payload).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {bearer_token}",
    }
    req = urlrequest.Request(function_url, data=body, headers=headers, method="POST")
    with urlrequest.urlopen(req, timeout=120) as res:
        raw = res.read().decode("utf-8")
    return json.loads(raw)


def extract_global_total_posts(sound_info: dict[str, Any]) -> int:
    candidates = [
        (((sound_info.get("musicInfo") or {}).get("stats") or {}).get("videoCount")),
        (((sound_info.get("musicInfo") or {}).get("music") or {}).get("videoCount")),
        (((sound_info.get("musicInfo") or {}).get("music") or {}).get("stats") or {}).get("videoCount"),
        ((sound_info.get("stats") or {}).get("videoCount")),
        ((sound_info.get("stats") or {}).get("video_count")),
    ]
    for value in candidates:
        total = to_int(value)
        if total > 0:
            return total
    return 0


def extract_creator_followers(video_data: dict[str, Any], author: dict[str, Any]) -> int:
    author_stats = video_data.get("authorStats", {})
    if not isinstance(author_stats, dict):
        author_stats = {}

    user_info = video_data.get("authorInfo", {})
    if not isinstance(user_info, dict):
        user_info = {}

    nested_stats = (user_info.get("stats") or {}) if isinstance(user_info, dict) else {}
    if not isinstance(nested_stats, dict):
        nested_stats = {}

    candidates = [
        author.get("followerCount"),
        author.get("follower_count"),
        author_stats.get("followerCount"),
        author_stats.get("follower_count"),
        nested_stats.get("followerCount"),
        nested_stats.get("follower_count"),
    ]
    for value in candidates:
        count = to_int(value)
        if count > 0:
            return count
    return 0


async def collect_videos(
    sound_ids: list[str],
    max_videos_per_sound: int,
    ms_token: str | None,
    browser: str,
    headless: bool,
    session_sleep_after_seconds: int,
    proxies: list[str] | None,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    output: list[dict[str, Any]] = []
    sound_stats: list[dict[str, Any]] = []

    async with TikTokApi() as api:
        session_kwargs: dict[str, Any] = {
            "num_sessions": 1,
            "sleep_after": max(1, session_sleep_after_seconds),
            "browser": browser,
            "headless": headless,
        }
        if ms_token:
            session_kwargs["ms_tokens"] = [ms_token]
        if proxies:
            session_kwargs["proxies"] = proxies

        try:
            await api.create_sessions(**session_kwargs)
        except TypeError:
            # Compatibility fallback for older TikTokApi signatures.
            fallback_kwargs: dict[str, Any] = {
                "num_sessions": 1,
                "sleep_after": max(1, session_sleep_after_seconds),
            }
            if ms_token:
                fallback_kwargs["ms_tokens"] = [ms_token]
            await api.create_sessions(**fallback_kwargs)

        for sound_id in sound_ids:
            sound = api.sound(id=sound_id)
            try:
                info = await sound.info()
                total_posts_global = extract_global_total_posts(info if isinstance(info, dict) else {})
            except Exception:
                total_posts_global = 0
            sound_stats.append({"sound_id": sound_id, "total_posts_global": total_posts_global})
            found = 0

            async for video in sound.videos(count=max_videos_per_sound):
                data = getattr(video, "as_dict", {}) or {}
                author = data.get("author", {}) if isinstance(data.get("author"), dict) else {}
                stats = data.get("stats", {}) if isinstance(data.get("stats"), dict) else {}

                video_id = str(data.get("id") or "")
                if not video_id:
                    continue

                output.append(
                    {
                        "video_id": video_id,
                        "sound_id": sound_id,
                        "video_url": f"https://www.tiktok.com/@{author.get('uniqueId', 'user')}/video/{video_id}",
                        "title": data.get("desc", ""),
                        "thumbnail_url": (data.get("video", {}) or {}).get("cover", ""),
                        "creator_username": author.get("uniqueId", ""),
                        "creator_followers": extract_creator_followers(data, author),
                        "posted_at": to_iso(data.get("createTime")),
                        "views": to_int(stats.get("playCount", 0)),
                        "likes": to_int(stats.get("diggCount", 0)),
                        "comments": to_int(stats.get("commentCount", 0)),
                        "shares": to_int(stats.get("shareCount", 0)),
                        "favorites": to_int(stats.get("collectCount", 0)),
                    }
                )

                found += 1
                if found >= max_videos_per_sound:
                    break

    return output, sound_stats


async def collect_videos_with_retries(
    sound_ids: list[str],
    max_videos_per_sound: int,
    ms_token: str | None,
    browser: str,
    headless: bool,
    session_sleep_after_seconds: int,
    proxies: list[str] | None,
    retries: int,
    retry_delay_seconds: int,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    last_error: Exception | None = None
    attempts = max(1, retries)
    strategy_profiles: list[tuple[str, bool]] = []
    for profile in [
        (browser, headless),
        ("webkit", True),
        ("chromium", True),
    ]:
        if profile not in strategy_profiles:
            strategy_profiles.append(profile)
    for attempt in range(1, attempts + 1):
        selected_browser, selected_headless = strategy_profiles[(attempt - 1) % len(strategy_profiles)]
        try:
            print(
                f"Fetch attempt {attempt}/{attempts} using browser={selected_browser}, headless={selected_headless}",
                file=sys.stderr,
            )
            return await collect_videos(
                sound_ids=sound_ids,
                max_videos_per_sound=max_videos_per_sound,
                ms_token=ms_token,
                browser=selected_browser,
                headless=selected_headless,
                session_sleep_after_seconds=session_sleep_after_seconds,
                proxies=proxies,
            )
        except Exception as error:  # pragma: no cover - runtime resilience path
            last_error = error if isinstance(error, Exception) else Exception(str(error))
            if attempt >= attempts:
                break
            delay = max(1, retry_delay_seconds) * attempt
            print(
                f"Fetch attempt {attempt}/{attempts} failed: {last_error}. Retrying in {delay}s...",
                file=sys.stderr,
            )
            await asyncio.sleep(delay)
    if last_error is not None:
        raise last_error
    raise RuntimeError("Fetch failed without explicit error")


async def main() -> int:
    supabase_url = (os.getenv("TTVM_SUPABASE_URL", "") or "").strip()
    anon_key = (os.getenv("TTVM_SUPABASE_ANON_KEY", "") or "").strip()
    sound_ids = parse_sound_ids((os.getenv("TTVM_TIKTOK_SOUND_IDS", "") or "").strip())
    function_name = (os.getenv("TTVM_FUNCTION_NAME", "tt-auto-fetch") or "").strip()
    ms_token = (os.getenv("TTVM_TIKTOK_MS_TOKEN", "") or "").strip() or None
    browser = (os.getenv("TTVM_TIKTOK_BROWSER", "webkit") or "").strip() or "webkit"
    headless = to_bool(os.getenv("TTVM_TIKTOK_HEADLESS", "false"), default=False)
    session_sleep_after_seconds = to_int(os.getenv("TTVM_SESSION_SLEEP_AFTER_SECONDS", "5"))
    proxy = (os.getenv("TTVM_TIKTOK_PROXY", "") or "").strip() or None
    proxies = [proxy] if proxy else None
    max_videos = to_int(os.getenv("TTVM_MAX_VIDEOS_PER_SOUND", "20"))
    fetch_retries = to_int(os.getenv("TTVM_FETCH_RETRIES", "3"))
    retry_delay_seconds = to_int(os.getenv("TTVM_RETRY_DELAY_SECONDS", "3"))

    if not supabase_url or not anon_key or not sound_ids:
        print(
            "Missing env. Required: TTVM_SUPABASE_URL, TTVM_SUPABASE_ANON_KEY, TTVM_TIKTOK_SOUND_IDS",
            file=sys.stderr,
        )
        return 1
    if not proxies:
        print(
            "Warning: TTVM_TIKTOK_PROXY is not set. GitHub-hosted runners are often rate-limited or blocked by TikTok.",
            file=sys.stderr,
        )

    function_url = f"{supabase_url}/functions/v1/{function_name}"
    print(f"Fetching TikTok videos for sounds: {', '.join(sound_ids)}")
    videos, sound_stats = await collect_videos_with_retries(
        sound_ids=sound_ids,
        max_videos_per_sound=max_videos,
        ms_token=ms_token,
        browser=browser,
        headless=headless,
        session_sleep_after_seconds=session_sleep_after_seconds,
        proxies=proxies,
        retries=fetch_retries,
        retry_delay_seconds=retry_delay_seconds,
    )

    if not videos:
        print("No videos collected. Try adding TTVM_TIKTOK_MS_TOKEN.", file=sys.stderr)
        return 2

    payload = {"sound_ids": sound_ids, "videos": videos, "sound_stats": sound_stats}
    print(f"Posting {len(videos)} videos to {function_url}")
    response = post_to_function(function_url, anon_key, payload)
    print(json.dumps(response, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: scripts/tt_try_fetch.ts
import { chromium, webkit, type Browser, type BrowserContext, type Page } from "playwright";

type Json = Record<string, unknown>;

type VideoPayload = {
  video_id: string;
  sound_id: string;
  video_url: string;
  title: string;
  thumbnail_url: string;
  creator_username: string;
  creator_followers: number;
  posted_at: string;
  views: number;
  likes: number;
  comments: number;
  shares: number;
  favorites: number;
};

type SoundStatPayload = {
  sound_id: string;
  total_posts_global: number;
};

const parseSoundIds = (raw: string) =>
  raw
    .split(",")
    .map((part) => part.trim())
    .filter((part) => part.length > 0);

const toInt = (value: unknown): number => {
  if (typeof value === "number" && Number.isFinite(value)) return Math.trunc(value);
  if (typeof value === "string") {
    const parsed = Number(value.replace(/,/g, "").trim());
    return Number.isFinite(parsed) ? Math.trunc(parsed) : 0;
  }
  return 0;
};

const toBool = (value: string | undefined, fallback: boolean): boolean => {
  if (!value) return fallback;
  const normalized = value.trim().toLowerCase();
  if (["1", "true", "yes", "y", "on"].includes(normalized)) return true;
  if (["0", "false", "no", "n", "off"].includes(normalized)) return false;
  return fallback;
};

const toIso = (value: unknown): string => {
  if (typeof value === "number" && Number.isFinite(value)) {
    return new Date(value * 1000).toISOString();
  }
  if (typeof value === "string" && value.trim()) {
    const direct = new Date(value);
    if (!Number.isNaN(direct.getTime())) return direct.toISOString();
    const asNumber = Number(value);
    if (Number.isFinite(asNumber)) return new Date(asNumber * 1000).toISOString();
  }
  return new Date().toISOString();
};

const firstNonEmptyString = (values: unknown[]): string => {
  for (const value of values) {
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return "";
};

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const postToFunction = async (functionUrl: string, bearerToken: string, payload: Json) => {
  const response = await fetch(functionUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${bearerToken}`,
    },
    body: JSON.stringify(payload),
  });

  const text = await response.text();
  if (!response.ok) {
    throw new Error(`Function request failed (${response.status}): ${text}`);
  }
  return text ? (JSON.parse(text) as Json) : {};
};

const readJsonScript = async (page: Page, scriptId: string): Promise<Json | null> => {
  const raw = await page
    .$eval(`#${scriptId}`, (node) => node.textContent ?? "")
    .catch(() => "");
  if (!raw.trim()) return null;
  try {
    return JSON.parse(raw) as Json;
  } catch {
    return null;
  }
};

const findObjectWithKey = (input: unknown, key: string): Json | null => {
  if (!input || typeof input !== "object") return null;
  const value = input as Json;
  if (key in value && value[key] && typeof value[key] === "object") {
    return value[key] as Json;
  }
  for (const child of Object.values(value)) {
    const found = findObjectWithKey(child, key);
    if (found) return found;
  }
  return null;
};

const extractFromSigiState = (
  sigi: Json,
  soundId: string,
  maxVideosPerSound: number,
): { videos: VideoPayload[]; totalPostsGlobal: number } => {
  const musicModule = (sigi.MusicModule as Json | undefined) ?? {};
  const itemModule = (sigi.ItemModule as Json | undefined) ?? {};

  const musicFromMap = (musicModule[soundId] as Json | undefined) ?? null;
  const musicFromAny =
    Object.values(musicModule).find((value) => {
      if (!value || typeof value !== "object") return false;
      const row = value as Json;
      return String(row.id ?? row.musicId ?? "") === soundId;
    }) ?? null;
  const music = ((musicFromMap ?? musicFromAny) as Json | null) ?? {};

  const totalPostsGlobal =
    toInt((music.stats as Json | undefined)?.videoCount) ||
    toInt((music.stats as Json | undefined)?.video_count) ||
    toInt((music.videoCount as unknown) ?? 0);

  const videos: VideoPayload[] = [];
  for (const row of Object.values(itemModule)) {
    if (!row || typeof row !== "object") continue;
    const item = row as Json;
    const musicRef = (item.music as Json | undefined) ?? {};
    const itemSoundId = firstNonEmptyString([
      musicRef.id,
      musicRef.mid,
      musicRef.musicId,
      item.musicId,
      item.sound_id,
    ]);
    if (itemSoundId !== soundId) continue;

    const id = firstNonEmptyString([item.id, item.aweme_id]);
    if (!id) continue;

    const author = (item.author as Json | undefined) ?? {};
    const stats = (item.stats as Json | undefined) ?? {};
    const video = (item.video as Json | undefined) ?? {};
    const authorStats = (item.authorStats as Json | undefined) ?? {};

    const username = firstNonEmptyString([author.uniqueId, author.unique_id, author.nickname, "user"]);
    videos.push({
      video_id: id,
      sound_id: soundId,
      video_url: `https://www.tiktok.com/@${username}/video/${id}`,
      title: firstNonEmptyString([item.desc, item.title]),
      thumbnail_url: firstNonEmptyString([video.cover, video.originCover, video.dynamicCover]),
      creator_username: username === "user" ? "" : username,
      creator_followers: toInt(author.followerCount) || toInt(authorStats.followerCount),
      posted_at: toIso(item.createTime),
      views: toInt(stats.playCount),
      likes: toInt(stats.diggCount),
      comments: toInt(stats.commentCount),
      shares: toInt(stats.shareCount),
      favorites: toInt(stats.collectCount),
    });

    if (videos.length >= maxVideosPerSound) break;
  }

  return { videos, totalPostsGlobal };
};

const extractFromUniversalData = (
  universal: Json,
  soundId: string,
  maxVideosPerSound: number,
): { videos: VideoPayload[]; totalPostsGlobal: number } => {
  const itemModule = findObjectWithKey(universal, "ItemModule") ?? {};
  const musicModule = findObjectWithKey(universal, "MusicModule") ?? {};

  const pseudoSigi: Json = {
    ItemModule: itemModule,
    MusicModule: musicModule,
  };
  return extractFromSigiState(pseudoSigi, soundId, maxVideosPerSound);
};

const collectForSound = async (
  page: Page,
  soundId: string,
  maxVideosPerSound: number,
): Promise<{ videos: VideoPayload[]; totalPostsGlobal: number }> => {
  const urlCandidates = [
    `https://www.tiktok.com/music/-${soundId}`,
    `https://www.tiktok.com/music/original-sound-${soundId}`,
  ];

  for (const url of urlCandidates) {
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45_000 }).catch(() => null);
    await page.waitForTimeout(2_000);

    const sigi = await readJsonScript(page, "SIGI_STATE");
    if (sigi) {
      const extracted = extractFromSigiState(sigi, soundId, maxVideosPerSound);
      if (extracted.videos.length > 0 || extracted.totalPostsGlobal > 0) return extracted;
    }

    const universal = await readJsonScript(page, "__UNIVERSAL_DATA_FOR_REHYDRATION__");
    if (universal) {
      const extracted = extractFromUniversalData(universal, soundId, maxVideosPerSound);
      if (extracted.videos.length > 0 || extracted.totalPostsGlobal > 0) return extracted;
    }
  }

  return { videos: [], totalPostsGlobal: 0 };
};

const withBrowser = async <T>(
  browserName: "chromium" | "webkit",
  headless: boolean,
  proxy: string | null,
  msToken: string | null,
  run: (page: Page) => Promise<T>,
) => {
  const browserLauncher = browserName === "webkit" ? webkit : chromium;
  const browser = await browserLauncher.launch({ headless });
  let context: BrowserContext | null = null;
  try {
    context = await browser.newContext(
      proxy
        ? {
            proxy: { server: proxy },
          }
        : undefined,
    );
    if (msToken) {
      await context.addCookies([
        {
          name: "msToken",
          value: msToken,
          domain: ".tiktok.com",
          path: "/",
          httpOnly: false,
          secure: true,
          sameSite: "None",
        },
      ]);
    }
    const page = await context.newPage();
    await page.setExtraHTTPHeaders({
      "user-agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
    });
    return await run(page);
  } finally {
    if (context) await context.close().catch(() => undefined);
    await browser.close().catch(() => undefined);
  }
};

const collectWithRetries = async (
  soundIds: string[],
  maxVideosPerSound: number,
  msToken: string | null,
  browser: "chromium" | "webkit",
  headless: boolean,
  proxy: string | null,
  retries: number,
  retryDelaySeconds: number,
) => {
  let lastError: unknown;
  const attempts = Math.max(1, retries);
  const strategy: Array<{ browser: "chromium" | "webkit"; headless: boolean }> = [];
  for (const profile of [
    { browser, headless },
    { browser: "webkit" as const, headless: true },
    { browser: "chromium" as const, headless: true },
  ]) {
    if (!strategy.some((row) => row.browser === profile.browser && row.headless === profile.headless)) {
      strategy.push(profile);
    }
  }

  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    const selected = strategy[(attempt - 1) % strategy.length];
    try {
      console.error(
        `Fetch attempt ${attempt}/${attempts} using browser=${selected.browser}, headless=${selected.headless}`,
      );

      const result = await withBrowser(
        selected.browser,
        selected.headless,
        proxy,
        msToken,
        async (page): Promise<{ videos: VideoPayload[]; soundStats: SoundStatPayload[] }> => {
          const allVideos: VideoPayload[] = [];
          const soundStats: SoundStatPayload[] = [];

          for (const soundId of soundIds) {
            const { videos, totalPostsGlobal } = await collectForSound(page, soundId, maxVideosPerSound);
            soundStats.push({ sound_id: soundId, total_posts_global: totalPostsGlobal });
            allVideos.push(...videos.slice(0, maxVideosPerSound));
          }

          return { videos: allVideos, soundStats };
        },
      );

      return result;
    } catch (error) {
      lastError = error;
      if (attempt >= attempts) break;
      const delay = Math.max(1, retryDelaySeconds) * attempt * 1000;
      console.error(`Fetch attempt ${attempt}/${attempts} failed: ${String(error)}. Retrying in ${delay / 1000}s...`);
      await sleep(delay);
    }
  }

  throw lastError instanceof Error ? lastError : new Error(String(lastError));
};

const main = async () => {
  const supabaseUrl = (process.env.TTVM_SUPABASE_URL ?? "").trim();
  const anonKey = (process.env.TTVM_SUPABASE_ANON_KEY ?? "").trim();
  const soundIds = parseSoundIds((process.env.TTVM_TIKTOK_SOUND_IDS ?? "").trim());
  const functionName = (process.env.TTVM_FUNCTION_NAME ?? "tt-auto-fetch").trim();
  const msToken = (process.env.TTVM_TIKTOK_MS_TOKEN ?? "").trim() || null;
  const proxy = (process.env.TTVM_TIKTOK_PROXY ?? "").trim() || null;
  const browserEnv = (process.env.TTVM_TIKTOK_BROWSER ?? "webkit").trim().toLowerCase();
  const browser: "chromium" | "webkit" = browserEnv === "chromium" ? "chromium" : "webkit";
  const headless = toBool(process.env.TTVM_TIKTOK_HEADLESS, true);
  const maxVideosPerSound = toInt(process.env.TTVM_MAX_VIDEOS_PER_SOUND ?? "20");
  const retries = toInt(process.env.TTVM_FETCH_RETRIES ?? "3");
  const retryDelaySeconds = toInt(process.env.TTVM_RETRY_DELAY_SECONDS ?? "3");

  if (!supabaseUrl || !anonKey || soundIds.length === 0) {
    console.error(
      "Missing env. Required: TTVM_SUPABASE_URL, TTVM_SUPABASE_ANON_KEY, TTVM_TIKTOK_SOUND_IDS",
    );
    process.exit(1);
  }
  if (!proxy) {
    console.error(
      "Warning: TTVM_TIKTOK_PROXY is not set. Some hosts may be rate-limited or blocked by TikTok.",
    );
  }

  const functionUrl = `${supabaseUrl}/functions/v1/${functionName}`;
  console.log(`Fetching TikTok videos for sounds: ${soundIds.join(", ")}`);

  const { videos, soundStats } = await collectWithRetries(
    soundIds,
    Math.max(1, maxVideosPerSound),
    msToken,
    browser,
    headless,
    proxy,
    Math.max(1, retries),
    Math.max(1, retryDelaySeconds),
  );

  if (videos.length === 0) {
    console.error("No videos collected. Try rotating msToken or adding TTVM_TIKTOK_PROXY.");
    process.exit(2);
  }

  const payload: Json = {
    sound_ids: soundIds,
    videos,
    sound_stats: soundStats,
  };
  console.log(`Posting ${videos.length} videos to ${functionUrl}`);
  const response = await postToFunction(functionUrl, anonKey, payload);
  console.log(JSON.stringify(response, null, 2));
};

void main().catch((error) => {
  console.error(error instanceof Error ? error.stack ?? error.message : String(error));
  process.exit(1);
});
